### PR TITLE
feat: helper support for unbalanced pool

### PR DESCRIPTION
**Pull request by fedgiac**
These changes make the helper support two-token pools with arbitrary weights.

More discussion on the math can be found [here](https://cowservices.slack.com/archives/C035TDPHQAJ/p1719567592184189?thread_ts=1719410711.653979&cid=C035TDPHQAJ) (I hope this link works for your team!). Message content copied below for accessibility.

More details are available as inline comments.

Notably, the `GetTradeableOrder` library is not needed anymore, since the math is now specific to this AMM version.

### How to test

Unit/integration tests have been changed to account for the new pool format. Logic changes are needed because the definition of the spot price isn't just the rate of the two pool balances when the weights are different.

Note that the "unreasonable amount" check has been removed. This is because it's hard to succinctly come up with a reasonableness check that works for both equal-weight and distinct-weight pool and is menaningfully readable. I don't think this is an issue since in both tests involving this check we also verify that the final price of the AMM coincides with `EXPECTED_FINAL_SPOT_PRICE`, which is reasonable by construction.

Some steps on CI are still expected to fail because:
- `secrets.MAINNET_RPC` doesn't appear to have been set (see [here](https://github.com/fedgiac/balancer-cow-amm/actions/runs/10769039841/job/29860298059) to confirm that the integration tests work)
- A Foundry update has changed the behavior of the default formatter. Fixing formatting would require changing many unrelated contracts, so I kept the old formatting in this PR.

<details><summary>Message with rationale for the math.</summary>

you want to calculate the sell amount so that, if the AMM sells tokens at the oracle price then its spot price equals the oracle price. So, if the oracle price is p, and you start with `b_x/w_x>p b_y/w_y`, you want to sell x for x/p such that

```
(b_x-x)/w_x=p (b_y+x/p)/w_y .
```

To put it another way, the calculations are correct if, in the end, the AMM spot price equals the Oracle. Also, the calculations have nothing to do with the "buy amount."

To conclude the calculations: you can solve the equation and get:

```
x = (w_x * w_y/(w_x+w_y))*(b_x/w_x - P (b_y/w_y))
```

note that the formula as in the reference implementation is

```
x = (b_x/2 - P (b_y/2))
```

so if you want to recycle it you need to normalize `w_x+w_y=2`  and redefine the pool reserves as  `b_x*w_y`  and `b_y*w_x`.
Then, once the calculations for the sell amount are correct, you just put its value in the `inGivenOut` formula to get the "buyAmount"

</details>

### CODE REVIEW

File: test/integration/BCoWHelper.t.sol [hunk 1]
```diff
@@ -41,9 +41,10 @@ contract BCoWHelperIntegrationTest is Test {
   IERC20 private constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
 
   uint256 constant TEN_PERCENT = 0.1 ether;
+  uint256 constant ONE_IN_A_BILLION = 0.000000001 ether;
   // NOTE: 1 ETH = 1000 DAI
-  uint256 constant INITIAL_DAI_BALANCE = 1000 ether;
-  uint256 constant INITIAL_WETH_BALANCE = 1 ether;
+  uint256 constant EQUAL_WEIGTH_INITIAL_DAI_BALANCE = 1000 ether;
+  uint256 constant EQUAL_WEIGTH_INITIAL_WETH_BALANCE = 1 ether;
```
Comment (mfw78) [inline test/integration/BCoWHelper.t.sol]:
```suggestion
  uint256 constant EQUAL_WEIGHT_INITIAL_DAI_BALANCE = 1000 ether;
  uint256 constant EQUAL_WEIGHT_INITIAL_WETH_BALANCE = 1 ether;
```

File: test/integration/BCoWHelper.t.sol [hunk 2]
```diff
@@ -69,14 +70,14 @@ contract BCoWHelperIntegrationTest is Test {
 
     DAI.approve(address(basicPool), type(uint256).max);
     WETH.approve(address(basicPool), type(uint256).max);
-    basicPool.bind(address(DAI), INITIAL_DAI_BALANCE, 4.2e18); // no weight
-    basicPool.bind(address(WETH), INITIAL_WETH_BALANCE, 4.2e18); // no weight
+    basicPool.bind(address(DAI), EQUAL_WEIGTH_INITIAL_DAI_BALANCE, 4.2e18); // no weight
+    basicPool.bind(address(WETH), EQUAL_WEIGTH_INITIAL_WETH_BALANCE, 4.2e18); // no weight
```
Comment (mfw78) [inline test/integration/BCoWHelper.t.sol]:
```suggestion
    basicPool.bind(address(DAI), EQUAL_WEIGHT_INITIAL_DAI_BALANCE, 4.2e18); // no weight
    basicPool.bind(address(WETH), EQUAL_WEIGHT_INITIAL_WETH_BALANCE, 4.2e18); // no weight
```

File: test/integration/BCoWHelper.t.sol [hunk 3]
```diff
@@ -69,14 +70,14 @@ contract BCoWHelperIntegrationTest is Test {
...
 
     DAI.approve(address(weightedPool), type(uint256).max);
     WETH.approve(address(weightedPool), type(uint256).max);
     // NOTE: pool is 80-20 DAI-WETH, has 4xDAI balance than basic, same spot price
-    weightedPool.bind(address(DAI), 4 * INITIAL_DAI_BALANCE, 8e18); // 80% weight
-    weightedPool.bind(address(WETH), INITIAL_WETH_BALANCE, 2e18); // 20% weight
+    weightedPool.bind(address(DAI), 4 * EQUAL_WEIGTH_INITIAL_DAI_BALANCE, 8e18); // 80% weight
+    weightedPool.bind(address(WETH), EQUAL_WEIGTH_INITIAL_WETH_BALANCE, 2e18); // 20% weight
```
Comment (mfw78) [inline test/integration/BCoWHelper.t.sol]:
```suggestion
    weightedPool.bind(address(DAI), 4 * EQUAL_WEIGHT_INITIAL_DAI_BALANCE, 8e18); // 80% weight
    weightedPool.bind(address(WETH), EQUAL_WEIGHT_INITIAL_WETH_BALANCE, 2e18); // 20% weight
```

File: test/integration/BCoWHelper.t.sol [hunk 4]
```diff
@@ -125,8 +128,8 @@ contract BCoWHelperIntegrationTest is Test {
     // if the first token is DAI and the second WETH then a price of 3000
     // DAI per WETH means a price vector of [1, 3000] (if the decimals are
     // different, as in WETH/USDC, then the atom amount is what counts).
-    prices[daiIndex] = INITIAL_WETH_BALANCE;
-    prices[wethIndex] = INITIAL_DAI_BALANCE * SKEWENESS_RATIO / 100;
+    prices[daiIndex] = EQUAL_WEIGTH_INITIAL_WETH_BALANCE;
+    prices[wethIndex] = EQUAL_WEIGTH_INITIAL_DAI_BALANCE * SKEWENESS_RATIO / 100;
```
Comment (mfw78) [inline test/integration/BCoWHelper.t.sol]:
```suggestion
    prices[daiIndex] = EQUAL_WEIGHT_INITIAL_WETH_BALANCE;
    prices[wethIndex] = EQUAL_WEIGHT_INITIAL_DAI_BALANCE * SKEWENESS_RATIO / 100;
```

**Review by mfw78 — COMMENTED** [review]
Reviewed for solidity standards, non-math logic errors, typos etc.

File: src/contracts/BCoWHelper.sol [hunk 5]
```diff
@@ -117,9 +97,88 @@ contract BCoWHelper is ICOWAMMPoolHelper, BMath {
...
+    uint256 balanceOutTimesWeightIn = bmul(reservesOut.balance, reservesIn.normWeight);
+    uint256 balanceInTimesWeightOut = bmul(reservesIn.balance, reservesOut.normWeight);
+    // This check compares the (weight-adjusted) pool spot price with the input
+    // price.
+    if (bmul(balanceInTimesWeightOut, priceNumerator) > bmul(balanceOutTimesWeightIn, priceDenominator)) {
+      (reservesOut, reservesIn) = (reservesIn, reservesOut);
+      (balanceOutTimesWeightIn, balanceInTimesWeightOut) = (balanceInTimesWeightOut, balanceOutTimesWeightIn);
+      (priceNumerator, priceDenominator) = (priceDenominator, priceNumerator);
     }
+    // The out amount is computed according to the following formula:
```
Comment (danielmkm) [inline src/contracts/BCoWHelper.sol]: The code (lines 134-135) does not align with the formula presented.

amountOut = bO * wI - p * bI * wO

File: src/contracts/BCoWHelper.sol [hunk 6]
```diff
@@ -117,9 +97,88 @@ contract BCoWHelper is ICOWAMMPoolHelper, BMath {
...
+    // aO = amountOut
+    // bI = reservesIn.balance                   bO * wI - p * bO * wI
+    // bO = reservesOut.balance            aO =  ---------------------
+    // wI = reservesIn.denormWeight                     wI + wO
```
Comment (danielmkm) [inline src/contracts/BCoWHelper.sol]: Can you add some comments to clarify that because `wI + wO = 1`, it can be excluded from the calculation below.

File: src/contracts/BCoWHelper.sol [hunk 7]
```diff
@@ -117,9 +97,88 @@ contract BCoWHelper is ICOWAMMPoolHelper, BMath {
...
+    // The price of this function is expressed as amount of token1 per amount
+    // of token0. The `prices` vector is expressed the other way around, as
+    // confirmed by dimensional analysis of the expression above.
+    uint256 priceNumerator = prices[1]; // x token = sell token = out amount
+    uint256 priceDenominator = prices[0];
+    uint256 balanceOutTimesWeightIn = bmul(reservesOut.balance, reservesIn.normWeight);
+    uint256 balanceInTimesWeightOut = bmul(reservesIn.balance, reservesOut.normWeight);
+    // This check compares the (weight-adjusted) pool spot price with the input
+    // price.
+    if (bmul(balanceInTimesWeightOut, priceNumerator) > bmul(balanceOutTimesWeightIn, priceDenominator)) {
```
Comment (danielmkm) [inline src/contracts/BCoWHelper.sol:145]: Can you explain the derivation here as it is hard to understand without context.

IMG_7197

File: src/contracts/BCoWHelper.sol (hunk 5 above)
Comment (fedgiac) [reply src/contracts/BCoWHelper.sol]: Addressed in https://github.com/balancer/cow-amm/pull/4/commits/1cec00ba5c50a62b227a3236057f4491179c61da . Does it work now?

File: src/contracts/BCoWHelper.sol (hunk 6 above)
Comment (fedgiac) [reply src/contracts/BCoWHelper.sol]: Addressed in https://github.com/balancer/cow-amm/pull/4/commits/026b61216b5c520454cc357aab05ecdd24012959 .

File: src/contracts/BCoWHelper.sol (hunk 7 above)
Comment (fedgiac) [reply src/contracts/BCoWHelper.sol:145]: Addressed in https://github.com/balancer/cow-amm/pull/4/commits/9b70c6e0d7168814f80687198accbb204ea1b0bc . Please let me know if I should phrase it differently.

File: src/contracts/BCoWHelper.sol [hunk 8]
```diff
@@ -117,9 +97,97 @@ contract BCoWHelper is ICOWAMMPoolHelper, BMath {
...
+    // The out amount is computed according to the following formula:
+    // aO = amountOut
+    // bI = reservesIn.balance                   bO * wI - p * bI * wO
+    // bO = reservesOut.balance            aO =  ---------------------
+    // wI = reservesIn.denormWeight                     wI + wO
+    // wO = reservesOut.denormWeight
+    // p  = priceNumerator / priceDenominator
+    //
+    // Note that in the code we use normalized weights instead of computing the
+    // full expression from raw weights. This is equivalent to assuming that
```
Comment (danielmkm) [inline src/contracts/BCoWHelper.sol:121]:
```suggestion
    // full expression from raw weights. Since BCoW pools only support two tokens, this is equivalent to assuming that
```

File: src/contracts/BCoWHelper.sol [hunk 9]
```diff
@@ -117,9 +97,97 @@ contract BCoWHelper is ICOWAMMPoolHelper, BMath {
...
+    // wI + wO = 1.
```
Comment (danielmkm) [inline src/contracts/BCoWHelper.sol]:
```suggestion
    // full expression from raw weights. Since BCoW pools support only two
    // tokens, this is equivalent to assuming that wI + wO = 1.
```

File: src/contracts/BCoWHelper.sol (hunk 7 above)
Comment (danielmkm) [reply src/contracts/BCoWHelper.sol:145]: Looks a lot better. The point we still see missing is that this check is deciding which token the pool should sell, but the comment doesn't really speak to that.

Comment (fedgiac) [reply src/contracts/BCoWHelper.sol:145]: Fair point, I updated the comment in https://github.com/balancer/cow-amm/pull/4/commits/7241d5fde9b872916a2e201262ea7098a3d595db . Does it make the role of this inequality clear?

Comment (danielmkm) [reply src/contracts/BCoWHelper.sol:145]: Looks great now!

**Review by danielmkm — APPROVED** [review]
Thanks for making the comment updates @fedgiac!
